### PR TITLE
feat: Add support for using static discovery documents

### LINKED ISSUE #1063: Convert to a static/packaged library
The Python discovery client library `google-api-python-client` currently downloads the discovery doc for a service at runtime and constructs a client. Convert the library to a static/packaged library. See the [Node client](https://github.com/googleapis/google-api-nodejs-client/tree/master/src) for an example of this.

The Python client has a [discovery_cache](https://github.com/googleapis/google-api-python-client/tree/master/googleapiclient/discovery_cache) that is only usable with the deprecated auth client (oauth2client) or App Engine Standard's memcache.  One possible solution is adjusting this cache to reference downloaded json instead.

### PR DESCRIPTION
**Pull request by parthea**
- Add new parameter `static_discovery` to `googleapiclient.discovery.build()`. When `static_discovery` is set to `True` only static discovery documents included in the library will be used.
- If you prefer to not use static discovery documents, you can disable this feature by setting the `static_discovery` argument of `googleapiclient.discovery.build()` to `False`. For example :

```
from googleapiclient.discovery import build
build("youtubereporting", "v1", static_discovery=False)
```

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-api-python-client/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1063 🦕

### CODE REVIEW

**Review by bryanlaura736 — APPROVED** [review]

File: googleapiclient/discovery_cache/__init__.py [hunk 1]
```diff
@@ -47,3 +48,36 @@ def autodetect():
...
+        A string containing the contents of the JSON discovery document,
+        otherwise None if the JSON discovery document was not found.
+    """
+
+    doc_name = None
+    content = None
+
+    # Extract the {apiVersion} and {api} from the uri which are the 2nd and 3rd
+    # last parts of the uri respectively.
+    # https://www.googleapis.com/discovery/v1/apis/{api}/{apiVersion}/rest
```
Comment (busunkim96) [inline googleapiclient/discovery_cache/__init__.py:70]: There are two different URI patterns (see `V1_DISCOVERY_URI` and `V2_DISCOVERY_URI` in `discovery.py`). Perhaps the service name and version can be passed from `build()` to `_retrieve_discovery_doc()` so the uri doesn't have to be parsed here?

File: googleapiclient/discovery_cache/__init__.py [hunk 2]
```diff
@@ -47,3 +48,36 @@ def autodetect():
     except Exception as e:
         LOGGER.warning(e, exc_info=True)
         return None
+
+def get_static_doc(uri):
+    """Retrieves the discovery document from the directory defined in
+    DISCOVERY_DOC_STATIC_DIR corresponding to the uri provided.
```
Comment (busunkim96) [inline googleapiclient/discovery_cache/__init__.py:54]: Variable above is named `DISCOVERY_DOC_DIR`

File: googleapiclient/discovery_cache/__init__.py (hunk 1 above)
Comment (parthea) [reply googleapiclient/discovery_cache/__init__.py]: Good catch! Fixed.

File: googleapiclient/discovery_cache/__init__.py (hunk 2 above)
Comment (parthea) [reply googleapiclient/discovery_cache/__init__.py]: Good catch! Fixed.

**Review by bryanlaura736 — APPROVED** [review]

**Review by danoscarmike — APPROVED** [review]

**Review by busunkim96 — APPROVED** [review]

**Comment by parthea** [comment]
@busunkim96 @SurferJeffAtGoogle I've updated the [docs](https://github.com/googleapis/google-api-python-client/pull/1109/files?short_path=298e509#diff-298e509e87a2e9501cb8f763fb811a38b5a10fb14ab30f4ef55ce7f0c58821b3), please can you take another look?

File: docs/start.md [hunk 3]
```diff
@@ -65,6 +65,17 @@ with build('drive', 'v3') as service:
...
...h is only supported with `oauth2client<4.0`. If the artifact is not available in the legacy cache and the `static_discovery` argument of `build()` is set to `True`, the library will attempt to retieve the discovery artifact from the static copy of the discovery artifacts that are shipped with the library. As a last resort if the discovery artifact is not on disk, the discovery artifact will be retrieved from the internet. The maintainers recommend keeping `static_discovery=True`, which is the default, in order to avoid reliability issues related to transient network errors. The library currently does not detect breaking changes of discovery artifacts, and as a result, user code is not guaranteed to continue to succeed or fail with different versions of `google-api-python-client`. Users can set `static_discovery=False` as workaround if there is an incompatible discovery artifact shipped with the library. In addition, users may roll back the version of `google-api-python-client` if there is a breaking change in a discovery artifact. The maintainers of `google-api-python-client` encourage users to report issues with discovery artifacts so that the discovery artifacts can be corrected.
```
Comment (busunkim96) [inline docs/start.md:68]: Thank you for adding this! It reads very nicely! 👍 

 One concern - `The library currently does not detect breaking changes of discovery artifacts` seems unsatisfying.  Users will have a hard time figuring out what changed between versions. NodeJS lists the added/deleted keys in the commit. https://github.com/googleapis/google-api-nodejs-client/commit/92b94ae7a3ace22248950961c6367eff2b3c15ff https://github.com/googleapis/google-api-nodejs-client/blob/master/CHANGELOG.md Can we do something similar in a follow up PR? @SurferJeffAtGoogle @JustinBeckwith

**Review by JustinBeckwith — APPROVED** [review]

**Review by bryanlaura736 — APPROVED** [review]

**Review by busunkim96 — APPROVED** [review]

File: docs/start.md (hunk 3 above)
Comment (parthea) [reply docs/start.md]: I agree. I've created #1154 to track this concern.

File: describe.py [hunk 4]
```diff
@@ -395,19 +395,23 @@ def document_api(name, version, uri):
...
     try:
         service = build(name, version)
+
+        http = build_http()
+        response, content = http.request(
+            uri or uritemplate.expand(
+                FLAGS.discovery_uri_template, {"api": name, "apiVersion": version}
+            )
+        )
+        if 'status' in response:
```
Comment (SurferJeffAtGoogle) [inline describe.py:405]: Why this change?

File: docs/start.md [hunk 5]
```diff
@@ -65,6 +65,8 @@ with build('drive', 'v3') as service:
...
...t of `build()` is set to `True`, the library will attempt to retrieve the discovery artifact from the legacy cache which is only supported with `oauth2client<4.0`. If the artifact is not available in the legacy cache and the `static_discovery` argument of `build()` is set to `True`, the library will attempt to retieve the discovery artifact from the static copy of the discovery artifacts that are shipped with the library. The maintainers recommend keeping `static_discovery=True`, which is the default, in order to avoid reliability issues related to transient network errors. The library currently does not detect breaking changes of discovery artifacts, and as a result, user code is not guaranteed to continue to succeed or fail with different versions of `google-api-python-client`. Users can set `static_discovery=False` as workaround if there is an incompatible discovery artifact shipped with the library. In addition, users may roll back the version of `google-api-python-client` if there is a breaking change in a discovery artifact. The maintainers of `google-api-python-client` encourage users to report issues with discovery artifacts so that the discovery artifacts can be corrected.
```
Comment (SurferJeffAtGoogle) [inline docs/start.md:68]: Could we describe this more in terms of the users' perspectives?  Something like the following

If reliability is most important, then set static_discovery=True.  It will use the service definition shipped in the library.
If always using the most recent version of the library is most important, then ...

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: describe.py (hunk 4 above)
Comment (parthea) [reply describe.py]: Good catch! This change isn't necessary. `describe.py` is only used to generate the docs located [here](https://github.com/googleapis/google-api-python-client/tree/master/docs/dyn). After taking another look at function `document_api`, it should not be retrieving discovery artifacts over the internet so I removed that code. Please take another look.

File: docs/start.md (hunk 5 above)
Comment (parthea) [reply docs/start.md]: Done. I also removed the text about breaking change detection as that bug is captured in #1154 . Please can you take another look?

**Comment by parthea** [comment]
@busunkim96 @SurferJeffAtGoogle Please can you take another look?

**Review by busunkim96 — APPROVED** [review]

File: docs/start.md (hunk 5 above)
Comment (parthea) [reply docs/start.md]: In an effort to move this PR forward, I'm going to resolve this comment as the feedback has been addressed.

File: describe.py (hunk 4 above)
Comment (parthea) [reply describe.py]: In an effort to move this PR forward, I'm going to resolve this comment as the feedback has been addressed.
